###  Q1. Write a Python code to implement the KNN classifier algorithm on load_iris dataset in sklearn.datasets.

In [16]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

#Load the iris dataset
df=load_iris()

#Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(df.data, df.target,test_size=0.3,random_state=42)

# Create a KNN classifier with k=3
knn= KNeighborsClassifier(n_neighbors=3)

# Fit the classifier to the training data
knn.fit(X_train, y_train)

# Make predictions on the testing data
y_pred = knn.predict(X_test)

# Calculate the accuracy of the classifier
accuracy = accuracy_score(y_test, y_pred)

# Print the accuracy
print("Accuracy:", accuracy)

Accuracy: 1.0


###  Q2. Write a Python code to implement the KNN regressor algorithm on load_boston dataset in sklearn.datasets.

In [48]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error

#Load the Boston Housing dataset
data_url = "http://lib.stat.cmu.edu/datasets/boston"
raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
target = raw_df.values[1::2, 2]

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(data, target, test_size=0.3, random_state=42)

# Create a KNN regressor with k=5
knn = KNeighborsRegressor(n_neighbors=5)

# Fit the regressor to the training data
knn.fit(X_train, y_train)

# Make predictions on the testing data
y_pred = knn.predict(X_test)

# Calculate the mean squared error of the regressor
mse = mean_squared_error(y_test, y_pred)

# Print the mean squared error
print("Mean Squared Error:", mse)

Mean Squared Error: 30.94554736842105


###  Q3. Write a Python code snippet to find the optimal value of K for the KNN classifier algorithm using cross-validation on load_iris dataset in sklearn.datasets.

In [50]:
from sklearn.datasets import load_iris
from sklearn.model_selection import cross_val_score
from sklearn.neighbors import KNeighborsClassifier

#Load the iris dataset
df=load_iris()

# Create a list of possible k values
k_values = list(range(1, 31))

# Create an empty list to store the cross-validation scores
cv_scores = []

# Perform 10-fold cross-validation for each value of k
for k in k_values:
    knn = KNeighborsClassifier(n_neighbors=k)
    scores = cross_val_score(knn, df.data, df.target, cv=10, scoring='accuracy')
    cv_scores.append(scores.mean())

# Find the optimal value of k with the highest cross-validation score
optimal_k = k_values[cv_scores.index(max(cv_scores))]
print("Optimal k value:", optimal_k)

Optimal k value: 13


###  Q4. Implement the KNN regressor algorithm with feature scaling on load_boston dataset in sklearn.datasets.

In [51]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler

#Load the Boston Housing dataset
data_url = "http://lib.stat.cmu.edu/datasets/boston"
raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
target = raw_df.values[1::2, 2]

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(data, target, test_size=0.3, random_state=42)

#Feature scaling
scaler=StandardScaler()
X_train=scaler.fit_transform(X_train)
X_test=scaler.transform(X_test)

# Create a KNN regressor with k=5
knn = KNeighborsRegressor(n_neighbors=5)

# Fit the regressor to the training data
knn.fit(X_train, y_train)

# Make predictions on the testing data
y_pred = knn.predict(X_test)

# Calculate the mean squared error of the regressor
mse = mean_squared_error(y_test, y_pred)

# Print the mean squared error
print("Mean Squared Error:", mse)

Mean Squared Error: 18.83503947368421


MSE decreased from `30.94` to `18.8` by using standard scaler.

###  Q5. Write a Python code snippet to implement the KNN classifier algorithm with weighted voting on load_iris dataset in sklearn.datasets.

In [57]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

# Load the iris dataset
iris = load_iris()

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.3)

# Create a KNN classifier object with weighted voting
knn = KNeighborsClassifier(n_neighbors=5, weights='distance')

# Train the classifier on the training data
knn.fit(X_train, y_train)

# Use the trained classifier to predict the labels of the test data
y_pred = knn.predict(X_test)

# Print the accuracy score of the classifier
print('Accuracy:', knn.score(X_test, y_test))

Accuracy: 0.9555555555555556


###  Q6. Implement a function to standardise the features before applying KNN classifier.

In [58]:
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

def knn_classifier(X_train, y_train, X_test, y_test, n_neighbors=5, weights='distance'):
    # Standardize the features
    scaler = StandardScaler()
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)
    
    # Create a KNN classifier object with weighted voting
    knn = KNeighborsClassifier(n_neighbors=n_neighbors, weights=weights)

    # Train the classifier on the training data
    knn.fit(X_train, y_train)

    # Use the trained classifier to predict the labels of the test data
    y_pred = knn.predict(X_test)

    # Print the accuracy score of the classifier
    accuracy = knn.score(X_test, y_test)
    print('Accuracy:', accuracy)
    
    return accuracy

###  Q7. Write a Python function to calculate the euclidean distance between two points.

In [59]:
import math
def euclidean_distance(point1, point2):
    distance = 0
    for i in range(len(point1)):
        distance += (point1[i] - point2[i])**2
    return math.sqrt(distance)

###  Q8. Write a Python function to calculate the manhattan distance between two points.

In [60]:
def manhattan_distance(point1, point2):
    distance = 0
    for i in range(len(point1)):
        distance += abs(point1[i] - point2[i])
    return distance